packages/firestore/src/util/bundle.ts

```diff
@@ -0,0 +1,215 @@
+    BundleElement | SizedBundleElement | null
+  > {
+    const lengthBuffer = await this.readLength();
+    if (lengthBuffer === null) {
+      return null;
+    }
+
+    const lengthString = this.textDecoder.decode(lengthBuffer);
+    const length = parseInt(lengthString, 10);
+    if (isNaN(length)) {
+      throw new Error(`length string (${lengthString}) is not valid number`);
+    }
+
+    const jsonString = await this.readJsonString(lengthBuffer.length, length);
+    // Update the internal buffer to drop the read length and json string.
+    this.buffer = this.buffer.slice(lengthBuffer.length + length);
+
+    if (!this.metadata) {
+      const element = JSON.parse(jsonString) as BundleElement;
+      return element;
+    } else {
+      return new SizedBundleElement(
+        JSON.parse(jsonString),
+        lengthBuffer.length + length
+      );
+    }
+  }
+
+  // First index of '{' from the underlying buffer.
+  private indexOfOpenBracket(): number {
+    return this.buffer.findIndex(v => v === 123);
+  }
+
+  // Reads from the beginning of the inernal buffer, until the first '{', and return
+  // the content.
+  // If reached end of the stream, returns a null.
+  private async readLength(): Promise<Uint8Array | null> {
+    let position = this.indexOfOpenBracket();
+    while (position < 0) {
+      const done = await this.pullMoreDataToBuffer();
+      if (done) {
+        if (this.buffer.length === 0) {
+          return null;
+        }
+        position = this.indexOfOpenBracket();
+        // Underlying stream is closed, and we still cannot find a '{'.
+        if (position < 0) {
+          throw new Error(
+            'Reach to the end of bundle when a length string is expected.'
+          );
+        }
+      } else {
+        position = this.indexOfOpenBracket();
+      }
+    }
+
+    return this.buffer.slice(0, position);
+  }
+
+  // Reads from a specified position from the internal buffer, for a specified
+  // number of bytes, pulling more data from the underlying stream if needed.
+  //
+  // Returns a string decoded from the read bytes.
+  private async readJsonString(start: number, length: number): Promise<string> {
+    while (this.buffer.length < start + length) {
+      const done = await this.pullMoreDataToBuffer();
+      if (done) {
+        throw new Error('Reach to the end of bundle when more is expected.');
+      }
+    }
+
+    return this.textDecoder.decode(this.buffer.slice(start, start + length));
+  }
+
+  // Pulls more data from underlying stream to internal buffer.
+  // Returns a boolean indicating whether the stream is finished.
+  private async pullMoreDataToBuffer(): Promise<boolean> {
+    const result = await this.reader.read();
+    if (!result.done) {
+      const newBuffer = new Uint8Array(
+        this.buffer.length + result.value.length
+      );
+      newBuffer.set(this.buffer);
+      newBuffer.set(result.value, this.buffer.length);
+      this.buffer = newBuffer;
```

Comment:
This is a very expensive way of consuming the buffer that creates `O(n^2)` copies. 

Can we not take advantage of the fact that we are using length-prefixed JSON and remove the internal buffer?

- Read first n bytes to get length (let's say n = 10)
- Length is 123, we read 7 bytes extra
- Read next 116 bytes, store in nextBytes
- JSON.parse(length.slice(7), nextBytes)
- Return as bundle element

Reply:
Note we cannot control how many bytes we read from `ReadableStream`, so an internal buffer cannot be avoided completely. 

Currently implementation only pulls more data when necessary, so the internal buffer is already kept as small as possible when there is a need to pull and concat.

Actually, I don't think it's `O(n^2)` copies in reality, because most of the data read from underlying stream is copied only once (into existing internal buffer), the part that will be copied again is the part that is the start of an element, which will be copied again when the next pull is called. 

The worst case is when one element requires multiple pulls, and each pulls will have to copy the old and new buffer. I feel this might not be very common though.